src/GraphQL/Validation/Rules.Custom/InputFieldsAndArgumentsOfCorrectLength.cs

```diff
@@ -0,0 +1,139 @@
+            {
+                _context = context;
+            }
+
+            public void VisitScalar(VariableDefinition variable, VariableName variableName, ScalarGraphType type, object variableValue, object parsedValue)
+            {
+
+            }
+
+            public void VisitList(VariableDefinition variable, VariableName variableName, ListGraphType type, object variableValue, object parsedValue)
+            {
+
+            }
+
+            public void VisitObject(VariableDefinition variable, VariableName variableName, IInputObjectGraphType type, object variableValue, object parsedValue)
+            {
+
+            }
+
+            public void VisitField(VariableDefinition variable, VariableName variableName, IInputObjectGraphType type, FieldType field, object variableValue, object parsedValue)
+            {
+                if (!field.HasAppliedDirectives())
+                    return;
+
+                var lengthDirective = field.GetAppliedDirectives().Find("length");
+                if (lengthDirective == null)
+                    return;
+
+                var min = lengthDirective.FindArgument("min")?.Value;
+                var max = lengthDirective.FindArgument("max")?.Value;
+
+                if (parsedValue == null)
+                {
+                    if (min != null)
+                        _context.ReportError(new InputFieldsAndArgumentsOfCorrectLengthError(_context, variable, variableName, null, (int?)min, (int?)max));
+                }
+                else if (parsedValue is string str)
+                {
+                    if (min != null && str.Length < (int)min || max != null && str.Length > (int)max)
+                        _context.ReportError(new InputFieldsAndArgumentsOfCorrectLengthError(_context, variable, variableName, str.Length, (int?)min, (int?)max));
+                }
+            }
+        }
+
+        /// <summary>
+        /// Returns a static instance of this validation rule.
+        /// </summary>
+        public static readonly InputFieldsAndArgumentsOfCorrectLength Instance = new InputFieldsAndArgumentsOfCorrectLength();
+
+        /// <inheritdoc/>
+        /// <exception cref="InputFieldsAndArgumentsOfCorrectLengthError"/>
+        public Task<INodeVisitor> ValidateAsync(ValidationContext context) => _nodeVisitor;
+
+        private static readonly Task<INodeVisitor> _nodeVisitor = new NodeVisitors(
+            new MatchingNodeVisitor<Document>((document, context) =>
+            {
+                var operation = string.IsNullOrWhiteSpace(context.OperationName)
+                   ? context.Document.Operations.FirstOrDefault()
+                   : context.Document.Operations.WithName(context.OperationName);
+
+                //TODO: change to get single variable / rework validation framework for variables
+                ExecutionHelper.GetVariableValues(context.Document, context.Schema, operation?.Variables, context.Inputs, new VariableVisitor(context));
```

Comment:
We need to allow validation rules so add variable visitors to a list within the validation context, which the document executer will pass to the variable parser.  There's probably a memory-optimized way to do that, but in any case, it should not run `GetVariableValues` here.

Reply:
@Shane32 Done. Today I was able to achieve this. The changes have had strong consequences. I have changed some of the tests. Along the way, I discovered erroneous tests. It's too late now, I'll do a PR tomorrow. We will need to discuss it.

Reply:
I updated PR.